kernel/src/vmm/execloop.rs

```diff
@@ -0,0 +1,146 @@
+        // (core_remap_ca() or core_create_vcpu()) so it's safe to use.
+        unsafe {
+            calling_area.write(caa_serviced)?;
+        }
+
+        Ok(caa.call_pending())
+    } else {
+        Ok(false)
+    }
+}
+
+fn get_svsm_request_message(vmsa_ref: &mut GuestVmsaRef) -> Option<GuestExitMessage> {
+    let vmsa = vmsa_ref.vmsa();
+
+    // Ignore guest exits that were not initiated by VMFEXIT.
+    if !matches!(vmsa.guest_exit_code, GuestVMExit::VMGEXIT) {
+        return None;
+    }
+    let rax = vmsa.rax;
+    let protocol = (rax >> 32) as u32;
+    let request = (rax & 0xffff_ffff) as u32;
+    let params = RequestParams::from_vmsa(vmsa);
+
+    match get_and_clear_caa_request_flag(vmsa_ref) {
+        Ok(pending) => {
+            if pending {
+                return Some(GuestExitMessage::Svsm((protocol, request, params)));
+            }
+        }
+        Err(SvsmReqError::RequestError(code)) => {
+            log::debug!(
+                "Soft error handling protocol {} request {}: {:?}",
+                protocol,
+                request,
+                code
+            );
+        }
+        Err(SvsmReqError::FatalError(err)) => {
+            panic!(
+                "Fatal error handling core protocol request {}: {:?}",
+                request, err
+            );
+        }
+    }
+
+    None
+}
+
+pub fn enter_guest(mut regs: &[GuestRegister]) -> GuestExitMessage {
+    let cpu = this_cpu();
+
+    // If no VMSA or CAA are configured, then the guest cannot be entered.
+    if cpu.update_guest_mappings().is_err() {
+        return GuestExitMessage::NoMappings;
+    }
+
+    // Set any guest registers that were supplied by the caller.
+
+    loop {
+        // No interrupts may be processed once guest APIC state is updated,
+        // since handling an interrupt may modify the guest APIC state
+        // calculations, which could cause state corruption.  If interrupts are
+        // disabled, then any additional guest APIC updates generated by the
+        // host will block the VMPL transition and permit reevaluation of
+        // guest APIC state.
+        let guard = {
+            let mut vmsa_ref = cpu.guest_vmsa_ref();
+            let caa_addr = vmsa_ref.caa_addr();
+            let vmsa = vmsa_ref.vmsa();
+
+            // Modify guest registers before disabling interrupts.
+            for reg in regs {
+                set_guest_register(vmsa, reg);
+            }
+            regs = &[];
+
+            // Interrupts must be disabled before updating APIC state.
+            let irq_guard = IrqGuard::new();
+
+            // Update APIC interrupt emulation state if required.
+            cpu.update_apic_emulation(vmsa, caa_addr);
+
+            // Make VMSA runnable again by setting EFER.SVME
+            vmsa.enable();
+
+            irq_guard
+        };
+
+        flush_tlb_global_sync();
+
+        switch_to_vmpl(GUEST_VMPL as u32);
+
+        drop(guard);
```

Comment:
Is this clearer? `let guard =` can even be omitted and just let `irq_guard` drop automatically as it goes out of scope.

```suggestion
        let mut vmsa_ref = cpu.guest_vmsa_ref();
        let caa_addr = vmsa_ref.caa_addr();
        let vmsa = vmsa_ref.vmsa();

        // Modify guest registers before disabling interrupts.
        for reg in regs {
            set_guest_register(vmsa, reg);
        }
        regs = &[];

        let guard = {
            // Interrupts must be disabled before updating APIC state.
            let irq_guard = IrqGuard::new();

            // Update APIC interrupt emulation state if required.
            cpu.update_apic_emulation(vmsa, caa_addr);

            // Make VMSA runnable again by setting EFER.SVME
            vmsa.enable();

            flush_tlb_global_sync();
            
            switch_to_vmpl(GUEST_VMPL as u32);
            
            irq_guard
        };
        drop(guard);
```

Reply:
@peterfang with your version, maybe we can avoid the explicit drop just not returning the guard, I mean something liker this:
```
        {
            // Interrupts must be disabled before updating APIC state.
            let irq_guard = IrqGuard::new();
            // Update APIC interrupt emulation state if required.
            cpu.update_apic_emulation(vmsa, caa_addr);
            // Make VMSA runnable again by setting EFER.SVME
            vmsa.enable();
            flush_tlb_global_sync();
            
            switch_to_vmpl(GUEST_VMPL as u32);
        }
```

Reply:
Peter's suggestion won't work because `vmsa_ref` must be dropped before calling `switch_to_vmpl()`.  I could clarify that in a comment, but that's the reason that `vmsa_ref` is captured within a separate block.  The challenge is that the IRQ guard must be established within that block but must outlive it, which is why it is passed out and dropped later.

I can look into explicitly dropping `vmsa_ref` without needing any lexical scope block, but if it's not trivial (which it may not be due to the number of other values that borrow from `vmsa_ref`), then keeping the scope block may actually be the simplest.